_posts/2019-03-26-gloo-onion-layers.md

```diff
@@ -0,0 +1,250 @@
+and which we didn't know yet, was *how* we were going design and expose these
+reusable bits.
+
+## Onion-Layered APIs
+
+I'm pleased to tell you that that after some collaborative discussion in issue
+threads, we've come up with a promising approach to designing Gloo APIs, and
+we've since formalized it a bit in `CONTRIBUTING.md`. I've nicknamed this
+approach "onion-layered" API design.
+
+Briefly, we want to build mid-level abstraction libraries on top of raw `-sys`
+bindings, build futures and streams integration on top of the mid-level APIs,
+and build high-level APIs on top of all that. But &mdash; crucially &mdash;
+every layer should be publicly exposed and reusable.
+
+While this approach to API design is certainly not novel, we want to very
+deliberately follow it so that we
+
+* maximize reusability for the larger ecosystem, and
+* exercise our mid-level APIs when building higher-level APIs, to ensure their
+  generality and suitability for acting as a solid foundation.
+
+As we go through and examine each layer, I'll use [the `setTimeout` and
+`setInterval` Web APIs][set-timeout] as a running example.
+
+## The Core: `wasm-bindgen`, `js-sys`, and `web-sys`
+
+The innermost layer are raw bindings built on top of [`wasm-bindgen`, `js-sys`
+and `web-sys`][announcing-web-sys]. These bindings are fast, have a light code
+size foot print, and are future-compatible with [the host bindings
+proposal][host-bindings].
+
+What they are *not* is super ergonomic to use all of the time. Using raw
+`web-sys` bindings directly can sometimes feel like making raw `libc` calls
+instead of leveraging Rust's nice `std` abstractions.
+
+Here is doing some operation after a 500 millisecond timeout using raw `web-sys`
+bindings:
+
+```rust
+use wasm_bindgen::{closure::Closure, JsCast};
+
+// Create a Rust `FnOnce` closure that is exposed to JavaScript.
+let closure = Closure::once(move || {
+    do_some_operation();
+});
+
+// Get the JavaScript function that reflects our Rust closure.
+let js_val = closure.as_ref();
+let js_func = js_val.unchecked_ref::<js_sys::Function>();
+
+// Finally, call the `window.setTimeout` API.
+let id = web_sys::window()
+    .expect("should have a `window`")
+    .set_timeout_with_callback_and_timeout_and_arguments_0(js_func, 500)
+    .expect("should set a timeout OK");
+
+// Then, if we ever decide we want to cancel the timeout, we do this:
+web_sys::window()
+    .expect("should have a `window`)
+    .clear_timeout_with_handle(timeout_id);
+```
+
+## The `callback` Layer
+
+When we look at the raw `web-sys` usage, there is a bit of type conversion
+noise, some unfortunate method names, and a handful of `unwrap`s for ignoring
+edge-case scenarios where we prefer to fail loudly rather than limp along. We
+can clean all these things up with the first of our "mid-level" API layers,
+which in the case of timers is the `callback` module in the `gloo_timers` crate
```

Comment:
Nit: do we have consensus on `callback` vs `callbacks`?

I think we have consensus on `futures`, so for consistency we should probably use `callbacks`?